keps/sig-node/2625-cpumanager-policies-thread-placement/README.md

```diff
@@ -0,0 +1,326 @@
+- When `smtalign` policy option is specified along with `static` CPU Manager policy, an additional check in the allocation logic of the `static` policy ensures that CPUs would be allocated such that full cores are allocated. Because of this check, a pod would never have to acquire single threads with the aim to fill partially-allocated cores.
+- In case request translates to partial occupancy of the cores, the Pod will not be admitted and would fail with `SMTAlignmentError`.
+
+
+
+### Resource Accounting
+
+To illustrate the behaviour of the `smtalign` policy option, we will consider the following CPU topology. We will use as example a CPU package with 16 physical cores, 2-way SMT-capable.
+
+![Example Topology](smtalign-topology.png)
+
+
+Let's consider a single container, requesting 5 isolated cores.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: frontend
+spec:
+  containers:
+  - name: app
+    image: images.my-company.example/app:v4
+    resources:
+      requests:
+        memory: "128Mi"
+        cpu: "5"
+      limits:
+        memory: "128Mi"
+        cpu: "5"
+```
+
+The `smtalign` policy option would need to make sure the remaining core on the half-allocated physical CPU is left unallocated to avoid noisy neighbours.
+
+![Example core allocation with the smtalign policy option when requesting a odd number of cores](smtalign-allocation-odd-cores.png)
+
+The container will then actually get more virtual cores (6) than what is requesting (5).
+
+| Requested (virtual) CPUs | Requested (physical) CPUs | Unallocatable (virtual) CPUs   | Total allocated (virtual) CPUs            |
+| ------------------------ | ------------------------- | ------------------------------ | ----------------------------------------- |
+| 5                        | 3                         | 1                              | 6                                         |
+| N                        | M = N / `threads_per_cpu` | X = N % `threads_per_cpu`      | `requested_vcpus` + `unallocatable_vcpus` |
+
+With `threads_per_cpu` is typical 2 on x86_64 with SMT enabled - but this number is not fixed and can change in future hardware implementation.
+
+In order to make the resource reporting consistent, and avoiding cascading changes in the system, we enforce the request constraints ad admission time.
+This approach follows what the Topology Manager already does.
+
+### Alternatives
+
+The only drawback of the proposed admission handler is that pods might have to overallocate resources.
+We evaluated possible alternatives to the extra admission control, but we eventually discarded all of them. We document them in this section.
```

Comment:
Another drawback is that pods that are rejected due to an `AdmissionError` do not get automatically rescheduled. You will need to run these workloads as a `deployment` or a `job` to ensure that they get rescheduled if they are rejected.

Reply:
Very fair point. Will add into the KEP text.